src/Microsoft.Extensions.HealthChecks/HealthCheckService.cs

```diff
@@ -0,0 +1,98 @@
+        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks, ILogger<HealthCheckService> logger)
+        {
+            Checks = healthChecks.ToDictionary(c => c.Name);
+            _logger = logger;
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                foreach (var check in Checks)
+                {
+                    _logger.LogDebug("Health check '{healthCheckName}' has been registered", check.Key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Runs all the health checks in the application and returns the aggregated status.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks</param>
+        /// <returns>
+        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
+        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
+        /// </returns>
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default)
+        {
+            var results = new Dictionary<string, HealthCheckResult>(Checks.Count);
+            foreach (var pair in Checks)
+            {
+                // If the health check does things like make Database queries using EF or backend HTTP calls,
+                // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
+                using (_logger.BeginScope(new HealthCheckLogScope(pair.Key)))
+                {
+                    HealthCheckResult result;
+                    try
+                    {
+                        _logger.LogTrace("Running health check: {healthCheckName}", pair.Key);
+                        result = await pair.Value.CheckStatusAsync(cancellationToken);
+                        _logger.LogTrace("Health check '{healthCheckName}' completed with status '{healthCheckStatus}'", pair.Key, result.Status);
+                    }
+                    catch (Exception ex)
+                    {
+                        // We don't log this as an error because a health check failing shouldn't bring down the active task.
+                        _logger.LogTrace(ex, "Health check '{healthCheckName}' threw an unexpected exception", pair.Key);
+                        result = new HealthCheckResult(HealthCheckStatus.Failed, ex, ex.Message, data: null);
+                    }
+
+                    // This can only happen if the result is default(HealthCheckResult)
+                    if (result.Status == HealthCheckStatus.Unknown)
+                    {
+                        // This is different from the case above. We throw here because a health check is doing something specifically incorrect.
+                        _logger.LogError("Health check '{healthCheckName}' returned a result with a status of Unknown", pair.Key);
```

Comment:
This will not become an Error in Application insights, which I don't think is fine. Need to new up the exception pass it into the log statement and then return it or fix the App Insights provider.

https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/f24a9d4d1d5471059c0d9e5b5edcff4925283837/src/Microsoft.ApplicationInsights.AspNetCore/Logging/Implementation/ApplicationInsightsLogger.cs#L50

Reply:
Yeah, makes sense. Non-Healthy Results are now allowed to hold Exceptions, so we have something to pass along (and can synthesize something if there is no exception).

Reply:
Hmm, wouldn't App Insights get the error coming out of the throw statement that follows this?